clippy_lints/src/transmute/transmute_ptr_to_ptr.rs

```diff
@@ -14,18 +15,49 @@ pub(super) fn check<'tcx>(
     from_ty: Ty<'tcx>,
     to_ty: Ty<'tcx>,
     arg: &'tcx Expr<'_>,
+    msrv: &Msrv,
 ) -> bool {
-    match (&from_ty.kind(), &to_ty.kind()) {
-        (ty::RawPtr(_, _), ty::RawPtr(to_ty, to_mutbl)) => {
+    match (from_ty.kind(), to_ty.kind()) {
+        (ty::RawPtr(from_pointee_ty, from_mutbl), ty::RawPtr(to_pointee_ty, to_mutbl)) => {
             span_lint_and_then(
                 cx,
                 TRANSMUTE_PTR_TO_PTR,
                 e.span,
                 "transmute from a pointer to a pointer",
                 |diag| {
                     if let Some(arg) = sugg::Sugg::hir_opt(cx, arg) {
-                        let sugg = arg.as_ty(Ty::new_ptr(cx.tcx, *to_ty, *to_mutbl));
-                        diag.span_suggestion(e.span, "try", sugg, Applicability::Unspecified);
+                        if from_mutbl == to_mutbl
+                            && to_pointee_ty.is_sized(cx.tcx, cx.param_env)
+                            && msrv.meets(msrvs::POINTER_CAST)
+                        {
+                            diag.span_suggestion_verbose(
+                                e.span,
+                                "use `pointer::cast` instead",
+                                format!("{}.cast::<{to_pointee_ty}>()", arg.maybe_par()),
+                                Applicability::MaybeIncorrect,
+                            );
+                        } else if from_pointee_ty == to_pointee_ty
```

Comment:
I think this could end up as a wrong suggestion for lifetime transmutes (`*mut Lt<'a>` to `*const Lt<'static>`). Maybe we can add a check if the from- or to type has `TypeFlags::RE_ERASED`?

Reply:
Interesting, is there a good example? I assume this would also effect ptr_as_ptr

Reply:
Can't think of a *good* (practical) example, but I was thinking of something like:
```rs
struct Lt<'a>(&'a ());

unsafe fn f(v: *mut Lt<'_>) -> *const Lt<'static> {
  std::mem::transmute(v)
}
```
This looks like it has .cast_const() for the suggestion, but that doesn't change the lifetime

Basically just this test case but with raw pointers and with differing mutability (so it doesn't go into the first `if` for pointer::cast):

https://github.com/rust-lang/rust-clippy/blob/afe811ad30198c966c6b69516c2a83f531c5c982/tests/ui/transmute_ptr_to_ptr.fixed#L10

Reply:
Thanks, unrealistic is fine I just can never remember which lifetimes are erased or early/late or any of that stuff 😅

Fixed it for `ptr_cast_constness` too